Allowed shortened grace period for pods in Kubelet

wzshiming opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind bug
/sig node

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

Fixes #83916
Fixes #84298
Fixes #87039
Fixes #88613
xref #98506
Fixes #100695

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Allowed shortened grace period for pods in Kubelet
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

smarterclayton commented:
I’d expect a test case to show the problem failing before (or an e2e if it’s simple to do via delete override).

jqmichael commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -636,7 +636,7 @@ func (m *kubeGenericRuntimeManager) killContainer(pod *v1.Pod, containerID kubec
 		gracePeriod = minimumGracePeriodInSeconds
 	}
-	if gracePeriodOverride != nil {
+	if gracePeriodOverride != nil && *gracePeriodOverride >= 0 {
 		gracePeriod = *gracePeriodOverride
 		klog.V(3).Infof("Killing container %q, but using a %d second grace period override", containerID, gracePeriod)
 	}
```
This is a nice addition. 

In practice, looks like `gracePeriodOverride` won't be set to negative value.

https://github.com/kubernetes/kubernetes/blob/acb43c7c4ac565eb2b51f98dcce32b5879383a19/pkg/kubelet/eviction/eviction_manager.go#L381-L384

wzshiming replied:
That's very helpful. Thank.

jqmichael reviewed (commented):
I think the issue isn't in the kubelet code, but on the server side. 

```
	if objectMeta.GetDeletionTimestamp() != nil {
		// if we are already being deleted, we may only shorten the deletion grace period
		// this means the object was gracefully deleted previously but deletionGracePeriodSeconds was not set,
		// so we force deletion immediately
		// IMPORTANT:
		// The deletion operation happens in two phases.
		// 1. Update to set DeletionGracePeriodSeconds and DeletionTimestamp
		// 2. Delete the object from storage.
		// If the update succeeds, but the delete fails (network error, internal storage error, etc.),
		// a resource was previously left in a state that was non-recoverable.  We
		// check if the existing stored resource has a grace period as 0 and if so
		// attempt to delete immediately in order to recover from this scenario.
		if objectMeta.GetDeletionGracePeriodSeconds() == nil || *objectMeta.GetDeletionGracePeriodSeconds() == 0 {
			return false, false, nil
		}
		// only a shorter grace period may be provided by a user
		if options.GracePeriodSeconds != nil {
			period := int64(*options.GracePeriodSeconds)
			if period >= *objectMeta.GetDeletionGracePeriodSeconds() {
				return false, true, nil
			}
```

https://github.com/kubernetes/apiserver/blame/d4c9a195921609cf81e3e950beaf246f934e0f4c/pkg/registry/rest/delete.go#L96-L110

As the commen suggest, in phase 1, the `DeletionGracePeriodSeconds` was set to a negative value (meaning `DeletionTimestamp` was **in the past**), and in phase 2, `options.GracePeriodSeconds` has to be set even lower, otherwise, it will be considered as pending graceful deletion and not performing the immediate delete. 

I think there're two possible fixes. 

1. Consider not allowing setting `DeletionTimestamp` in the past (`DeletionGracePeriodSeconds` should be non-negative). In that case, we could convert `options.GracePeriodSeconds` to 0 if negative. 
2. Accept that `DeletionTimestamp` could be set in the past. In that case, in phase 2, we treat negative value of `objectMeta.GetDeletionGracePeriodSeconds()` the same as 0, and allow immediate delete. 

Thoughts? 

/CC: @smarterclayton @liggitt @lavalamp @deads2k

lavalamp commented:
DeletionGracePeriodSeconds should not accept negative values, how did that happen?

ayberk commented:
@lavalamp We're still trying to figure it out, but current hypothesis is a 3rd party node termination handler manually set it to -1. There are no guards against negative values from what I can tell.

lavalamp commented:
We should update validation to forbid it.

liggitt commented:
> We should update validation to forbid it.

or clip to zero

maybe-related issues with weirdness around deletionTimestamp and deletionGracePeriodSeconds:
* https://github.com/kubernetes/kubernetes/issues/87039#issuecomment-572787482
* https://github.com/kubernetes/kubernetes/issues/83916#issuecomment-545994420

ayberk commented:
Clipping the zero would be my recommendation as well. The delete checks are always looking for `not nil && equals to 0`. Both clipping to zero or checking for `<= 0` instead of equality would work.

wzshiming commented:
> > We should update validation to forbid it.
> 
> or clip to zero, especially if there's existing persisted data containing negative values
> 
> maybe-related issues with weirdness around deletionTimestamp and deletionGracePeriodSeconds:
> 
> * [#87039 (comment)](https://github.com/kubernetes/kubernetes/issues/87039#issuecomment-572787482)
> * [#83916 (comment)](https://github.com/kubernetes/kubernetes/issues/83916#issuecomment-545994420)

It seems to be related, let me take a look at it as a whole

wzshiming commented:
Make sure this problem.

> Shortened grace period and deletionTimestamp do actually show up in the API, but the kubelet does not actually honor the shortened value in the API and continues allowing the pod to have the original grace period before terminating the process and doing the final pod deletion.

In the original kubelet implementation, if a pod is deleted, all subsequent updates will be ignored.
I added a delete operation that determines whether it is an earlier deletion, and if it is, execute this delete operation. but due to the implementation of cri, a new delete operation will re-emit a SIGTERM signal, for this, I added a FeatureGate to control whether to enable the ability to allow duplicate deletion.

wzshiming commented:
In the original, if a pod is deleting, all subsequent updates will be ignored. I added shortened grace period allowed.
but will re-emit a SIGTERM signal, so I added FeatureGate `ShortenedGracePeriod` to switch whether to enable this feature.

wzshiming commented:
/retest

wzshiming commented:
/retest

derekwaynecarr commented:
/assign

derekwaynecarr commented on test/e2e_node/shortened_grace_period_test.go:
```diff
@@ -0,0 +1,83 @@
+			f.PodClient().DeleteSync(podName, metav1.DeleteOptions{
+				GracePeriodSeconds: &gracePeriodZero,
+			}, 1*time.Second)
+			framework.ExpectEqual(time.Since(start) < 3*time.Second, true, "kubelet cannot update GracePeriodSeconds")
```
the `kubelet` isnt really the actor here, right?  the error message is slightly confusing as a result.

derekwaynecarr commented on pkg/kubelet/pod_workers.go:
```diff
@@ -194,38 +194,51 @@ func (p *podWorkers) managePodLoop(podUpdates <-chan UpdatePodOptions) {
+	} else if isShortenedGracePeriod(working, options) {
+		// The previous worker needs to wait until the pod is successfully deleted before executing the next one,
+		// we close the previous worker and start a new worker to perform the pod delete operation
+		p.removeWorker(pod.UID)
```
just noting that this is right to call versus `ForgetWorker` as we hold the pod lock.

derekwaynecarr reviewed (changes requested):
one minor nit, i want to also ensure there is no locks being held with runtime layers if they receive duplicate requests.

/cc @mrunalp can you check to ensure cri-o works with this patch?

derekwaynecarr commented:
I anticipate two kill pod operations will actually be in flight, and so just understanding that the second call is not blocked by any locking on the first call is good to validate.  I also am not sure if a spurious event could be thrown if/when one of those kill pod calls potentially errors.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -283,8 +296,34 @@ func (p *podWorkers) checkForUpdates(uid types.UID) {
+		return false
+	}
+	now := time.Now()
+	oldDeletionTimestamp := getDeleteTimestamp(old, now)
```
/hold

No component should EVER read deletion timestamp and act on it.

You may *only* act on the current time + deletionSeconds compared to when you saw the event the first time.

smarterclayton replied:
The logic of deleting a pod or shortening grace period should be:

1. When the kubelet sees a pod is deleted for the first time, it should take the current time (`time.Now()`) and record that as the time the first deletion was seen `t_delete`, and then trigger a container TERM (stopContainer) with an expected time to force terminate `t_kill = t_delete + max(grace_period, 3s)`.
2. pod stop should only be invoked one time on a pod as long as the kubelet has not restarted
3. If grace period is reduced after stopContainer is requested at least once, we must reduce the time subtracted from the original request (original grace period - current grace period) and execute the kill (SIGKILL, not TERM) to the container runtime, but make sure than no less than 3s have elapsed - so the updated `t_kill = t_delete + max(new_grace_period, 3s)`.  If `t_kill > t_now`, we should kill the pod immediately.

smarterclayton replied:
But it's really important that shortening deletion time should NEVER cause a second SIGTERM to be sent - it's the container runtimes responsibility to only send SIGTERM to a terminated process once, and then SIGKILL after the elapsed timeout has occurred.

derekwaynecarr replied:
I would expect the runtime to not dispatch TERM twice if it had already dispatched it once prior.

I would expect the runtime to just adjust its waiting period before sending SIGKILL.

This is why we need runtimes to ensure they do the right thing as I prefer kubelet stays simple.

/cc @mrunalp

mrunalp replied:
Yeah, we are working on a patch for this and should be able to ensure that only one SIGTERM is sent.

mikebrow replied:
Containerd will also need a patch to mirror the (new?) desire for only sending one SIGTERM when two CRI container stop requests are received for the same container. Concerned about possibility of regressions here.

dims commented:
I like the containerd PR better ... https://github.com/containerd/containerd/pull/5021

we should let this be done in the CRI implementations themselves!

wzshiming commented:
/retest

wzshiming commented:
> I like the containerd PR better ... [containerd/containerd#5021](https://github.com/containerd/containerd/pull/5021)
> 
> we should let this be done in the CRI implementations themselves!

@dims Thank.

This PR is a fix that cannot shorten the grace period.
https://github.com/containerd/containerd/pull/5021 This is to fix the signal will be sent repeatedly after shortening the grace period.
Both of these are required.

wzshiming commented:
/retest

pacoxu commented:
/kind bug

troy0820 commented:
/priority important-soon

SergeyKanzhelev commented:
/assign @ehashman

BenTheElder commented:
The CRI impl side of this is shipping in KIND jobs now.
It is not in the GCE jobs yet.

mrunalp commented:
Needs rebase

troy0820 commented:
/hold 
@wzshiming Please rebase PR.

troy0820 commented:
/hold cancel

wzshiming commented:
/retest

wzshiming commented:
/retest

wzshiming commented:
/retest

wzshiming commented:
/retest

wzshiming commented:
/retest

wzshiming commented:
/retest

wzshiming commented:
/retest

wzshiming commented:
/retest

wzshiming commented:
/retest

ehashman commented:
Please ensure tests are passing.

wzshiming commented:
/retest

wzshiming committed:
Add e2e test

Signed-off-by: Shiming Zhang <wzshiming@foxmail.com>

wzshiming commented:
/retest

wzshiming commented:
/retest

wzshiming commented:
@ehashman 

The test has been fixed, and containerd 1.5 (Include patch https://github.com/containerd/containerd/pull/5021) has been released. I think PR should be merged. Is there anything else that needs to be done?

wzshiming commented:
/retest

wzshiming commented:
/retest

wzshiming commented:
/retest

wzshiming commented:
PING @ehashman

smarterclayton commented:
> It looks like this PR #98507 fixes the same problem

Yes it's the same direction.  Note there are lots of other bugs in here so we'll have to tease apart what needs to be done - I don't want to miss the opportunity if we have gaps in the sync model.  The risk is changing how pod workers terminate needs more discussion than what has happened in previous attempts to fix it.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -193,42 +206,95 @@ func (p *podWorkers) managePodLoop(podUpdates <-chan UpdatePodOptions) {
 		update, found := p.lastUndeliveredWorkUpdate[pod.UID]
 		if !found || update.UpdateType != kubetypes.SyncPodKill {
 			p.lastUndeliveredWorkUpdate[pod.UID] = *options
+			p.recordDeletingTime(options, now)
```
SyncPodKill is not the only way that a pod can start terminating - I think we may need to consolidate the SyncPodKill and normal termination paths.

smarterclayton commented:
@ehashman is going to start a doc and share it, i tried to summarize briefly here: https://github.com/kubernetes/kubernetes/issues/102149#issuecomment-845266779

I want to hold this PR until we get a design for changes to pod worker, and then that should also help make this PR easier to fix.

ehashman commented:
Discussion doc here: https://docs.google.com/document/d/1DvAmqp9CV8i4_zYvNdDWZW-V0FMk1NbZ8BAOvLRD3Ds/edit#

Shared with k8s SIG Node mailing list

wzshiming commented:
/remove-lifecycle stale
/lifecycle frozen

wzshiming committed:
Shortened grace period allowed for pods in Kubelet

wzshiming commented:
Hi @smarterclayton

I tested in the latest version, the problem persists and is getting worse. At least it could be deleted with --force. Now even if it is forcibly deleted, the Pod is still in the CRI.
that syncTerminatingPod does not handle Context, can we bypass this problem first, and reduce the Pod leakage that this problem may bring.

adisky commented:
/cc

dims commented:
/milestone v1.24

jyotimahapatra commented:
Hi 👋 I'm checking in from the bug triage team for release 1.24. Is this PR targeted for release 1.24?

wzshiming commented:
@jyotimahapatra
Since PR #102344 was merged, both 1.22 and 1.23 exist, when Pod that is deleting cannot be forced deleted. Earlier versions, just can't shorten the grace period but can be forced to delete.

dims commented:
@ehashman @SergeyKanzhelev do we need this for v1.24?

pacoxu commented:
/remove-lifecycle stale
Does this fix https://github.com/kubernetes/kubernetes/issues/109352 as well?
